Auth0/CredentialsManager.swift

```diff
@@ -193,4 +192,44 @@ public struct CredentialsManager {
             }
         }
     }
+
+    func hasExpired(_ credentials: Credentials) -> Bool {
```

Comment:
Check logic here please, basically if either token has expired then it should return true

Reply:
Current logic looks correct (and tests should verify) ... but we could be much more clear here if we returned on the expired condition instead of setting variables:

```swift
if credentials.expiresIn <= Date() { 
    return true
}

var tokenDecoded = decode(jwt: credentials.idToken);
var exp = tokenDecoded["exp"] as? Double;

if Date(timeIntervalSince1970: exp) < Date() { 
    return true
}

return false
```

That's unlikely to be correct (let alone idiomatic) Swift 😆 but hopefully the message is clear.

Reply:
@cocojoe The `hasExpired` logic is correct, as josh pointed out. But I think the performance impact could be reduced a lot if instead of checking the exp date **of each token** and _decode_ the id token **every time** you call this method, you just save the exp date most close to expire at "save credentials" step, and then just have a single date comparison check on this function.

Reply:
Yeah I had similar logic to above but I think a mix up in tests sent me down the wrong path, have corrected now.

I don't mind the idea of a single expiry time based upon lowest. Will leave it for a future internal refactor though.